File: clientes/go/cliente.go
package main

import (
	"fmt"
	"log"
	"math/rand"
	"os"
	"strings"
	"sync"
	"time"

	"projeto-mensageria/shared/go/mensageria"
	"projeto-mensageria/shared/go/protos"

	"github.com/pebbe/zmq4"
	"google.golang.org/protobuf/proto"
)

const (
	orqEndpointDefault = "tcp://orquestrador:5555"
	proxySubEndpoint   = "tcp://proxy:5558"
	nomeUsuarioDefault = "cliente_go"
	nomeCanalDefault   = "canal_go"
)

func main() {
	orqEndpoint := getEnv("ORQ_ENDPOINT", orqEndpointDefault)
	subEndpoint := getEnv("PROXY_SUB_ENDPOINT", proxySubEndpoint)
	nomeUsuario := getEnv("CLIENTE_NOME", nomeUsuarioDefault)
	nomeCanal := getEnv("CLIENTE_CANAL", nomeCanalDefault)

	sock, err := zmq4.NewSocket(zmq4.DEALER)
	if err != nil {
		log.Fatalf("Novo socket: %v", err)
	}
	defer sock.Close()

	if err := sock.Connect(orqEndpoint); err != nil {
		log.Fatalf("Conectar ao orquestrador: %v", err)
	}
	log.Printf("[CLIENTE] Conectado ao orquestrador em %s", orqEndpoint)

	subSock, err := zmq4.NewSocket(zmq4.SUB)
	if err != nil {
		log.Fatalf("Novo socket SUB: %v", err)
	}
	defer subSock.Close()
	if err := subSock.Connect(subEndpoint); err != nil {
		log.Fatalf("Conectar SUB ao proxy: %v", err)
	}
	log.Printf("[CLIENTE] Conectado ao proxy sub em %s", subEndpoint)

	cliente := &Cliente{
		sock:       sock,
		subSock:    subSock,
		subscribed: map[string]bool{},
		clock:      &mensageria.RelogioProcesso{},
		origem:     mensageria.OrigemLabel(mensageria.RoleCliente),
	}

	for i := 0; i < 3; i++ {
		if cliente.fazerLogin(nomeUsuario) {
			break
		}
		time.Sleep(1 * time.Second)
	}

	cliente.iniciarReceptor()
	canais := cliente.listarCanais()
	if len(canais) < 5 {
		novoCanal := fmt.Sprintf("%s_%d", nomeCanal, rand.Intn(10000))
		cliente.criarCanal(novoCanal)
		canais = cliente.listarCanais()
	}
	for len(cliente.subscribed) < 3 && len(cliente.subscribed) < len(canais) {
		escolhido := canais[rand.Intn(len(canais))]
		cliente.inscrever(escolhido)
	}
	for {
		canais = cliente.listarCanais()
		if len(canais) == 0 {
			time.Sleep(1 * time.Second)
			continue
		}
		canal := canais[rand.Intn(len(canais))]
		for i := 0; i < 10; i++ {
			_ = cliente.publicar(canal, randomMensagem(12))
			time.Sleep(1 * time.Second)
		}
	}
}

type Cliente struct {
	sock       *zmq4.Socket
	subSock    *zmq4.Socket
	subscribed map[string]bool
	clock      *mensageria.RelogioProcesso
	origem     string
	mu         sync.Mutex
}

func getEnv(name, defaultVal string) string {
	if v := os.Getenv(name); v != "" {
		return v
	}
	return defaultVal
}

func agoraTexto() string {
	agora := time.Now().UnixNano()
	return fmt.Sprintf("%d.%09d", agora/1_000_000_000, agora%1_000_000_000)
}

func conteudoNome(env *protos.Envelope) string {
	switch env.GetConteudo().(type) {
	case *protos.Envelope_LoginReq:
		return "login_req"
	case *protos.Envelope_LoginRes:
		return "login_res"
	case *protos.Envelope_CreateChannelReq:
		return "create_channel_req"
	case *protos.Envelope_CreateChannelRes:
		return "create_channel_res"
	case *protos.Envelope_ListChannelsReq:
		return "list_channels_req"
	case *protos.Envelope_ListChannelsRes:
		return "list_channels_res"
	case *protos.Envelope_PublishReq:
		return "publish_req"
	case *protos.Envelope_PublishRes:
		return "publish_res"
	default:
		return "desconhecido"
	}
}

func (c *Cliente) enviarEAguardar(env *protos.Envelope) (*protos.Envelope, error) {
	data, err := mensageria.EnvelopeBytes(env)
	if err != nil {
		return nil, err
	}
	log.Printf("[CLIENTE] Enviando %s %s", conteudoNome(env), mensageria.CabecalhoTexto(env.GetCabecalho()))
	if _, err := c.sock.SendBytes(data, 0); err != nil {
		return nil, err
	}
	reply, err := c.sock.RecvBytes(0)
	if err != nil {
		return nil, err
	}
	respEnv, err := mensageria.EnvelopeFromBytes(reply)
	if err != nil {
		return nil, err
	}
	c.clock.OnReceive(respEnv.GetCabecalho().GetRelogioLogico())
	log.Printf("[CLIENTE] Recebido %T %s", respEnv.GetConteudo(), mensageria.CabecalhoTexto(respEnv.GetCabecalho()))
	return respEnv, nil
}

func (c *Cliente) fazerLogin(nomeUsuario string) bool {
	cab := mensageria.NovoCabecalho(c.origem, c.clock)
	req := &protos.LoginRequest{
		Cabecalho:   cab,
		NomeUsuario: nomeUsuario,
	}
	env := &protos.Envelope{
		Cabecalho: cab,
		Conteudo:  &protos.Envelope_LoginReq{LoginReq: req},
	}

	respEnv, err := c.enviarEAguardar(env)
	if err != nil {
		log.Printf("[CLIENTE] Erro ao enviar/aguardar login: %v", err)
		return false
	}

	_, ok := respEnv.GetConteudo().(*protos.Envelope_LoginRes)
	if !ok {
		log.Printf("[CLIENTE] Resposta inesperada ao login: %T", respEnv.GetConteudo())
		return false
	}
	res := respEnv.GetLoginRes()
	if res.GetStatus() == protos.Status_STATUS_SUCESSO {
		log.Printf("[CLIENTE] Login bem-sucedido para '%s'", nomeUsuario)
		return true
	}
	log.Printf("[CLIENTE] Falha no login: %s", res.GetErroMsg())
	return false
}

func (c *Cliente) criarCanal(nomeCanal string) {
	cab := mensageria.NovoCabecalho(c.origem, c.clock)
	req := &protos.CreateChannelRequest{
		Cabecalho: cab,
		NomeCanal: nomeCanal,
	}
	env := &protos.Envelope{
		Cabecalho: cab,
		Conteudo:  &protos.Envelope_CreateChannelReq{CreateChannelReq: req},
	}

	respEnv, err := c.enviarEAguardar(env)
	if err != nil {
		log.Printf("[CLIENTE] Erro ao criar canal: %v", err)
		return
	}
	_, ok := respEnv.GetConteudo().(*protos.Envelope_CreateChannelRes)
	if !ok {
		log.Printf("[CLIENTE] Resposta inesperada a criação de canal: %T", respEnv.GetConteudo())
		return
	}
	res := respEnv.GetCreateChannelRes()
	if res.GetStatus() == protos.Status_STATUS_SUCESSO {
		log.Printf("[CLIENTE] Canal '%s' criado com sucesso", nomeCanal)
	} else {
		log.Printf("[CLIENTE] Falha ao criar canal '%s': %s", nomeCanal, res.GetErroMsg())
	}
}

func (c *Cliente) listarCanais() []string {
	cab := mensageria.NovoCabecalho(c.origem, c.clock)
	req := &protos.ListChannelsRequest{Cabecalho: cab}
	env := &protos.Envelope{
		Cabecalho: cab,
		Conteudo:  &protos.Envelope_ListChannelsReq{ListChannelsReq: req},
	}

	respEnv, err := c.enviarEAguardar(env)
	if err != nil {
		log.Printf("[CLIENTE] Erro ao listar canais: %v", err)
		return nil
	}
	_, ok := respEnv.GetConteudo().(*protos.Envelope_ListChannelsRes)
	if !ok {
		log.Printf("[CLIENTE] Resposta inesperada a listagem de canais: %T", respEnv.GetConteudo())
		return nil
	}
	res := respEnv.GetListChannelsRes()
	canais := res.GetCanais()
	if len(canais) == 0 {
		log.Printf("[CLIENTE] Canais existentes: (nenhum)")
	} else {
		log.Printf("[CLIENTE] Canais existentes: %s", strings.Join(canais, ", "))
	}
	return canais
}

func (c *Cliente) publicar(canal, mensagem string) bool {
	cab := mensageria.NovoCabecalho(c.origem, c.clock)
	req := &protos.PublishRequest{Cabecalho: cab, Canal: canal, Mensagem: mensagem}
	env := &protos.Envelope{
		Cabecalho: cab,
		Conteudo:  &protos.Envelope_PublishReq{PublishReq: req},
	}
	respEnv, err := c.enviarEAguardar(env)
	if err != nil {
		log.Printf("[CLIENTE] Erro ao publicar: %v", err)
		return false
	}
	res, ok := respEnv.GetConteudo().(*protos.Envelope_PublishRes)
	if !ok {
		log.Printf("[CLIENTE] Resposta inesperada ao publicar: %T", respEnv.GetConteudo())
		return false
	}
	if res.PublishRes.GetStatus() != protos.Status_STATUS_SUCESSO {
		log.Printf("[CLIENTE] Falha ao publicar em '%s': %s", canal, res.PublishRes.GetErroMsg())
		return false
	}
	return true
}

func (c *Cliente) inscrever(canal string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.subscribed[canal] {
		return
	}
	if err := c.subSock.SetSubscribe(canal); err != nil {
		log.Printf("[CLIENTE] Erro ao inscrever no canal '%s': %v", canal, err)
		return
	}
	c.subscribed[canal] = true
	log.Printf("[CLIENTE] Inscrito no canal '%s'", canal)
}

func (c *Cliente) iniciarReceptor() {
	go func() {
		for {
			msg, err := c.subSock.RecvMessageBytes(0)
			if err != nil || len(msg) < 2 {
				time.Sleep(200 * time.Millisecond)
				continue
			}
			channelMsg := &protos.ChannelMessage{}
			err = proto.Unmarshal(msg[1], channelMsg)
			if err != nil {
				continue
			}
			c.clock.OnReceive(channelMsg.GetRelogioLogico())
			log.Printf(
				"[CLIENTE] [CANAL=%s] msg='%s' remetente=%s envio=%s recebimento=%s relogio=%d",
				string(msg[0]),
				channelMsg.GetMensagem(),
				channelMsg.GetRemetente(),
				mensageria.TimestampTexto(channelMsg.GetTimestampEnvio()),
				agoraTexto(),
				channelMsg.GetRelogioLogico(),
			)
		}
	}()
}

func randomMensagem(length int) string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, length)
	for i := range b {
		b[i] = chars[rand.Intn(len(chars))]
	}
	return string(b)
}

File: clientes/python/cliente.py
import logging
import os
import random
import string
import threading
import time
from typing import List

import zmq

from shared.mensageria import (
    RelogioProcesso,
    cabecalho_texto,
    envelope_bytes,
    envelope_from_bytes,
    novo_cabecalho,
    origem_label,
    timestamp_to_text,
)
from shared.protos import contrato_pb2


logging.basicConfig(level=logging.INFO, format="%(asctime)s [CLIENTE] %(message)s")


ORQ_ENDPOINT = os.getenv("ORQ_ENDPOINT", "tcp://orquestrador:5555")
PROXY_SUB_ENDPOINT = os.getenv("PROXY_SUB_ENDPOINT", "tcp://proxy:5558")
NOME_USUARIO = os.getenv("CLIENTE_NOME", "cliente_python")
NOME_CANAL = os.getenv("CLIENTE_CANAL", "canal_padrao")


def _agora_texto() -> str:
    agora_ns = time.time_ns()
    return f"{agora_ns // 1_000_000_000}.{agora_ns % 1_000_000_000:09d}"


class Cliente:
    def __init__(self) -> None:
        self.context = zmq.Context.instance()
        self.socket = self.context.socket(zmq.DEALER)
        self.socket.connect(ORQ_ENDPOINT)
        self.sub_socket = self.context.socket(zmq.SUB)
        self.sub_socket.connect(PROXY_SUB_ENDPOINT)
        self.subscribed_channels: set[str] = set()
        self.relogio = RelogioProcesso()
        self.origem = origem_label("cliente")
        logging.info("Conectado ao orquestrador em %s", ORQ_ENDPOINT)
        logging.info("Conectado ao proxy sub em %s", PROXY_SUB_ENDPOINT)

    def _enviar_e_aguardar(self, env: contrato_pb2.Envelope) -> contrato_pb2.Envelope:
        tipo = env.WhichOneof("conteudo") or "desconhecido"
        logging.info("Enviando %s %s", tipo, cabecalho_texto(env.cabecalho))
        self.socket.send(envelope_bytes(env))
        data = self.socket.recv()
        resp_env = envelope_from_bytes(data)
        self.relogio.on_receive(resp_env.cabecalho.relogio_logico)
        logging.info(
            "Recebido %s %s",
            resp_env.WhichOneof("conteudo") or "desconhecido",
            cabecalho_texto(resp_env.cabecalho),
        )
        return resp_env

    def fazer_login(self, nome_usuario: str) -> bool:
        env = contrato_pb2.Envelope()
        env.cabecalho.CopyFrom(novo_cabecalho(self.origem, self.relogio))
        login_req = contrato_pb2.LoginRequest()
        login_req.cabecalho.CopyFrom(env.cabecalho)
        login_req.nome_usuario = nome_usuario
        env.login_req.CopyFrom(login_req)

        resp_env = self._enviar_e_aguardar(env)
        conteudo = resp_env.WhichOneof("conteudo")
        if conteudo != "login_res":
            logging.error("Resposta inesperada ao login: %s", conteudo)
            return False

        res = resp_env.login_res
        if res.status == contrato_pb2.STATUS_SUCESSO:
            logging.info("Login bem-sucedido para '%s'", nome_usuario)
            return True

        logging.warning("Falha no login: %s", res.erro_msg)
        return False

    def criar_canal(self, nome_canal: str) -> None:
        env = contrato_pb2.Envelope()
        env.cabecalho.CopyFrom(novo_cabecalho(self.origem, self.relogio))
        req = contrato_pb2.CreateChannelRequest()
        req.cabecalho.CopyFrom(env.cabecalho)
        req.nome_canal = nome_canal
        env.create_channel_req.CopyFrom(req)

        resp_env = self._enviar_e_aguardar(env)
        conteudo = resp_env.WhichOneof("conteudo")
        if conteudo != "create_channel_res":
            logging.error("Resposta inesperada a criação de canal: %s", conteudo)
            return

        res = resp_env.create_channel_res
        if res.status == contrato_pb2.STATUS_SUCESSO:
            logging.info("Canal '%s' criado com sucesso", nome_canal)
        else:
            logging.warning("Falha ao criar canal '%s': %s", nome_canal, res.erro_msg)

    def listar_canais(self) -> List[str]:
        env = contrato_pb2.Envelope()
        env.cabecalho.CopyFrom(novo_cabecalho(self.origem, self.relogio))
        req = contrato_pb2.ListChannelsRequest()
        req.cabecalho.CopyFrom(env.cabecalho)
        env.list_channels_req.CopyFrom(req)

        resp_env = self._enviar_e_aguardar(env)
        conteudo = resp_env.WhichOneof("conteudo")
        if conteudo != "list_channels_res":
            logging.error("Resposta inesperada a listagem de canais: %s", conteudo)
            return []

        res = resp_env.list_channels_res
        canais = list(res.canais)
        logging.info("Canais existentes: %s", ", ".join(canais) if canais else "(nenhum)")
        return canais

    def publicar(self, canal: str, mensagem: str) -> bool:
        env = contrato_pb2.Envelope()
        env.cabecalho.CopyFrom(novo_cabecalho(self.origem, self.relogio))
        req = contrato_pb2.PublishRequest()
        req.cabecalho.CopyFrom(env.cabecalho)
        req.canal = canal
        req.mensagem = mensagem
        env.publish_req.CopyFrom(req)

        resp_env = self._enviar_e_aguardar(env)
        conteudo = resp_env.WhichOneof("conteudo")
        if conteudo != "publish_res":
            logging.error("Resposta inesperada ao publicar: %s", conteudo)
            return False
        res = resp_env.publish_res
        if res.status == contrato_pb2.STATUS_SUCESSO:
            return True
        logging.warning("Falha ao publicar em '%s': %s", canal, res.erro_msg)
        return False

    def inscrever(self, canal: str) -> None:
        if canal in self.subscribed_channels:
            return
        self.sub_socket.setsockopt(zmq.SUBSCRIBE, canal.encode("utf-8"))
        self.subscribed_channels.add(canal)
        logging.info("Inscrito no canal '%s'", canal)

    def iniciar_receptor(self) -> None:
        def _loop() -> None:
            while True:
                try:
                    topic, payload = self.sub_socket.recv_multipart()
                    msg = contrato_pb2.ChannelMessage()
                    msg.ParseFromString(payload)
                    self.relogio.on_receive(msg.relogio_logico)
                    logging.info(
                        "[CANAL=%s] msg='%s' remetente=%s envio=%s recebimento=%s relogio=%s",
                        topic.decode("utf-8"),
                        msg.mensagem,
                        msg.remetente,
                        timestamp_to_text(msg.timestamp_envio),
                        _agora_texto(),
                        msg.relogio_logico,
                    )
                except Exception as exc:
                    logging.warning("Erro no receptor de pub/sub: %s", exc)
                    time.sleep(0.2)

        thread = threading.Thread(target=_loop, daemon=True)
        thread.start()


def main() -> None:
    cliente = Cliente()

    for _ in range(3):
        if cliente.fazer_login(NOME_USUARIO):
            break
        time.sleep(1)

    cliente.iniciar_receptor()

    canais = cliente.listar_canais()
    if len(canais) < 5:
        sufixo = "".join(random.choices(string.ascii_lowercase + string.digits, k=4))
        novo_canal = f"{NOME_CANAL}_{sufixo}"
        cliente.criar_canal(novo_canal)
        canais = cliente.listar_canais()

    while len(cliente.subscribed_channels) < 3 and len(cliente.subscribed_channels) < len(canais):
        disponiveis = [c for c in canais if c not in cliente.subscribed_channels]
        if not disponiveis:
            break
        cliente.inscrever(random.choice(disponiveis))

    while True:
        canais = cliente.listar_canais()
        if not canais:
            time.sleep(1)
            continue
        canal_escolhido = random.choice(canais)
        for _ in range(10):
            msg = "".join(random.choices(string.ascii_lowercase + string.digits, k=12))
            cliente.publicar(canal_escolhido, msg)
            time.sleep(1)


if __name__ == "__main__":
    main()
